prometheusremotewrite: release lock when wal empty

tombrk opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->

Previously, the wal tailing routine held an exclusive lock over the entire wal, while waiting for new sections to be written.

This led to a deadlock situtation, as a locked wal naturally can not be written to.

To remedy this, the lock is now only held during actual read attempts, not in between.

Furthermore, inotify watching of the wal dir has been replaced with channel-based signaling between the producers and the tailing routine, so that the latter blocks until any new writes have happened.

**Link to tracking Issue:**

Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19363
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/15277


**Testing:**

An "end to end" test has been added where the WAL is first initialized without any initial data, so it goes into waiting mode. Writes are performed afterwards and the read-back result is verified.

This previously led to a deadlock and now no longer does.

It has also been confirmed that goroutines no longer build up and actual http requests are now being made.

tombrk committed:
prometheusremotewrite: release lock when wal empty

Previously, the wal tailing routine held an exclusive lock over the
entire wal, while waiting for new sections to be written.

This led to a deadlock situtation, as a locked wal naturally can not be written to.

To remedy this, the lock is now only held during actual read attempts,
not in between.

Furthermore, inotify watching of the wal dir has been replaced with
channel-based signaling between the producers and the tailing routine,
so that the latter blocks until any new writes have happened.

frzifus commented on exporter/prometheusremotewriteexporter/wal.go:
```diff
@@ -324,113 +332,81 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
 			return err
 		}
 		wIndex := prwe.wWALIndex.Add(1)
+		prwe.log.Debug("write", zap.Uint64("index", wIndex))
```
Do we still need it or can we set it to trace level? (in that case, it should be a bit more explicit)

sh0rez replied:
turns out – there is no `trace` level in zap ... so we need to either leave it as debug or remove entirely

frzifus replied:
how is the full log message displayed?

sh0rez replied:
```
=== RUN   TestWAL_E2E
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:364	read	{"index": 1}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:390	wal empty - waiting for write
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:335	write	{"index": 1}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:335	write	{"index": 2}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:335	write	{"index": 3}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:364	read	{"index": 1}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:364	read	{"index": 2}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:364	read	{"index": 3}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:364	read	{"index": 4}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:390	wal empty - waiting for write
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:364	read	{"index": 4}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:390	wal empty - waiting for write
```

frzifus commented on exporter/prometheusremotewriteexporter/wal.go:
```diff
@@ -324,113 +332,81 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
+		defer prwe.mu.Unlock()
 
-			return req, nil
+		prwe.log.Debug("read", zap.Uint64("index", index))
```
same here

frzifus commented on exporter/prometheusremotewriteexporter/wal.go:
```diff
@@ -324,113 +332,81 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
-			}()
+		req, err := try()
+		if errors.Is(err, wal.ErrNotFound) {
+			prwe.log.Debug("wal empty - waiting for write")
```
and here

frzifus commented:
Whats the state here? :) ping @sh0rez

codeboten commented:
Looks like the checks hung, re-opening to trigger the checks

codeboten reviewed (commented):
Please add a changelog entry for the bug fix

codeboten reviewed (commented):
Looks like a test is failing, please have a look

```
--- FAIL: TestWAL_persist (0.01s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa696b6]

goroutine 292 [running]:
testing.tRunner.func1.2({0x11fa360, 0x1abbe70})
	/opt/hostedtoolcache/go/1.20.4/x64/src/testing/testing.go:1526 +0x372
testing.tRunner.func1()
	/opt/hostedtoolcache/go/1.20.4/x64/src/testing/testing.go:1529 +0x650
panic({0x11fa360, 0x1abbe70})
	/opt/hostedtoolcache/go/1.20.4/x64/src/runtime/panic.go:890 +0x263
go.uber.org/zap.(*Logger).check(0x0, 0xff, {0x1306013, 0x5})
	/home/runner/go/pkg/mod/go.uber.org/zap@v1.24.0/logger.go:298 +0x76
go.uber.org/zap.(*Logger).Debug(0x149d400?, {0x1306013, 0x5}, {0xc0003cb2c0, 0x1, 0x1})
	/home/runner/go/pkg/mod/go.uber.org/zap@v1.24.0/logger.go:211 +0x55
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter.(*prweWAL).persistToWAL(0xc0000a60e0, {0xc000175e20, 0x2, 0x0?})
	/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter/wal.go:324 +0x308
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter.TestWAL_persist(0xc0001f4d00)
	/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter/wal_test.go:138 +0xb11
testing.tRunner(0xc0001f4d00, 0x135cf28)
	/opt/hostedtoolcache/go/1.20.4/x64/src/testing/testing.go:1576 +0x217
created by testing.(*T).Run
	/opt/hostedtoolcache/go/1.20.4/x64/src/testing/testing.go:1629 +0x806
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter	0.074s
FAIL
```

tombrk committed:
prometheusremotewrite: wal default nop-logger

Initializes the WAL with a default nop logger, instead of previous nil
which led to panics.

tombrk committed:
doc: add changelog entry

tombrk committed:
prometheusremotewrite: go mod tidy

tombrk committed:
*: resolve merge conflicts

sh0rez commented:
@codeboten any chance you could take another look?

frzifus commented:
ping @sh0rez

frzifus commented on exporter/prometheusremotewriteexporter/wal.go:
```diff
@@ -66,13 +68,17 @@ func newWAL(walConfig *WALConfig, exportSink func(context.Context, []*prompb.Wri
 		rWALIndex:  &atomic.Uint64{},
 		wWALIndex:  &atomic.Uint64{},
-	}, nil
+		log:        zap.NewNop(),
```
There is a logger available in the `TelemetrySettings`. But adding a `zap.NewNop()` logger here, should be equal to the state before without any logger, right?

Not that I have a strong opinion about it - I just would like to resolve it at some point :smile:

MovieStoreGuy replied:
I think it is more the point of, if we are adding a logger, we need some way of providing a configuration for it.

sh0rez replied:
the logger is taken from the context once `run()` is called, but we need a place to store it for long running operations.

not setting this var leads to a nil pointer that panics, so its `NewNop()` instead to avoid that. afaict this is never called before run() anways

frzifus reviewed (commented):
@sh0rez I created a small [PR](https://github.com/sh0rez/opentelemetry-collector-contrib/pull/1) to your branch to address the requested changes.

MovieStoreGuy commented on exporter/prometheusremotewriteexporter/wal.go:
```diff
@@ -313,113 +322,81 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
 
-			// Now increment the WAL's read index.
-			prwe.rWALIndex.Add(1)
+	try := func() (*prompb.WriteRequest, error) {
```
Could I ask that you move this to be a method as part of the type?

My concern being is that trying to unit test this means a lot of potential set up and keeping things simple and isolated helps checking where things break (or making sure they break correctly).

MovieStoreGuy commented on exporter/prometheusremotewriteexporter/wal.go:
```diff
@@ -313,113 +322,81 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-prwe.stopChan:
+			return nil, fmt.Errorf("attempt to read from WAL after stopped")
```
I would suggest using `errors.New` instead of `fmt.Errorf` since it should be "cheaper" to create the value since it doesn't need to parse the string for any formatting symbols.

Alternatively, since this is static you can define the error at the type of the file as part of the "static" vars.

MovieStoreGuy commented on exporter/prometheusremotewriteexporter/wal_test.go:
```diff
@@ -101,6 +103,7 @@ func TestWAL_persist(t *testing.T) {
 
 	pwal, err := newWAL(config, doNothingExportSink)
 	require.Nil(t, err)
+	pwal.log = zap.Must(zap.NewDevelopment())
```
To make your life easier, I suggest using `zaptest.NewLogger(t)` within tests so that you get logs only if the tests fails.

MovieStoreGuy commented on exporter/prometheusremotewriteexporter/wal_test.go:
```diff
@@ -156,3 +159,64 @@ func TestWAL_persist(t *testing.T) {
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	log, err := zap.NewDevelopment()
+	if err != nil {
+		panic(err)
+	}
```
You shouldn't need another logger here, I suggest you create one at the start of the test and use that throughout.

MovieStoreGuy commented on exporter/prometheusremotewriteexporter/wal_test.go:
```diff
@@ -156,3 +159,64 @@ func TestWAL_persist(t *testing.T) {
+	ctx = contextWithLogger(ctx, log)
+
+	if err := wal.run(ctx); err != nil {
+		panic(err)
```
Please use `require.NoError` instead of panicing.

MovieStoreGuy reviewed (commented):
Just a few things from me.

tombrk committed:
*: forward to main

tombrk committed:
*: review feedback

sh0rez commented:
@frzifus @Aneurysm9 @codeboten somehow dropped the ball on this, but it's still relevant.

rebased and addressed review comments, please take another look!

tombrk committed:
*: don't panic

jpkrohling reviewed (approved):
LGTM

frzifus reviewed (approved):
Thx, for picking this up again. :)

tombrk committed:
*: make linter happy

jpkrohling commented on exporter/prometheusremotewriteexporter/wal_test.go:
```diff
@@ -149,3 +152,57 @@ func TestWAL_persist(t *testing.T) {
+	}
+
+	// wait until the tail routine is no longer busy
+	wal.rNotify <- struct{}{}
```
Just to confirm: if I remove this, the test will be blocked on the `<-done` until it times out. That's the purpose of this test, and what you are fixing here, right?

Aneurysm9 replied:
I'm also a bit confused as to why this is needed here.  `run()` will fork and (indirectly) call `read()`, blocking until it returns, which will happen when a record is read from the WAL or the context is cancelled.  `read()` will try to read from the WAL, find nothing, and then block reading from `rNotify`.  `persistToWAL()` will then acquire the mutex, signal on `rNotify`, and write to the WAL.  Once `persistToWAL()` signals on `rNotify` then `read()` starts blocking on acquiring the mutex in `tryRead()`, eventually acquiring it once `persistToWAL()` executes its deferred unlock.  `read()` will then get a successful result from `tryRead()`, popping all the way back out to `continuallyPopWALThenExport()` which may or may not actually export (probably not with the default truncate frequency and buffer size), and then goes right back into `read()`.  At this point the WAL is empty and `tryRead()` returns the sentinel error `read()` uses to decide to wait on `rNotify`.  Since `persistToWAL()` has returned nothing is going to send anything on `rNotify` until this line here.  Sending a value here causes `read()` to try again, but nothing happens.  Then the context is cancelled and a bunch of `select` statements hit their `<-ctx.Done()` case, causing `read()` to return `nil, context.Canceled` to `continuallyPopWALThenExport()`, which returns `context.Canceled` to `run()`, executing its deferred export along the way, causing `sink` to close `done` and allowing this function to reach its requirement that the exported data equals the input data.

tl;dr: I don't see what writing to this channel does that isn't also accomplished by cancelling the context.

jpkrohling replied:
@sh0rez , are you still interested in pursuing this?

jpkrohling commented on exporter/prometheusremotewriteexporter/wal.go:
```diff
@@ -70,8 +72,12 @@ func newWAL(walConfig *WALConfig, exportSink func(context.Context, []*prompb.Wri
 		wWALIndex:  &atomic.Uint64{},
+
+		// populated from context in run()
+		log: zap.NewNop(),
```
I had the exact same feeling, but thought it would be ok to have it like that given it kind of follows the pattern of the existing code. I'd be ok to have this as part of a follow up PR.

jpkrohling replied:
I had the exact same feeling, but thought it would be ok to have it like that given it kind of follows the pattern of the existing code. I'd be ok to have this as part of a follow up PR.

Aneurysm9 replied:
I'm torn on this.  I feel strongly that we shouldn't be knowingly introducing new problems when we also know what the fix is, particularly when the fix isn't a big lift.  I also realize this is a purely internal change and doesn't expose any of the wrongness outside of the package.

The other conversation about the blocking on the notify channel in the test seems more important to me to resolve.  If we come to a resolution on that but this isn't fixed I think I'd be at a position where I wouldn't approve this PR but wouldn't object to it being merged with an understanding that this needs to be fixed as a fast follower.
